### PR TITLE
Update purchase pallet to FRAME v2

### PR DESCRIPTION
**Pull request by shaunxw**
Part of #2882 

Converts the `Purchase` pallet to the new pallet attribute macro introduced in #6877.

Following the upgrade guidelines here: https://crates.parity.io/frame_support/attr.pallet.html#upgrade-guidelines.

## ⚠️ Breaking Change ⚠️ 

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

> storages now use PalletInfo for module_prefix instead of the one given to decl_storage: Thus any use of this pallet in construct_runtime! should be careful to update name in order not to break storage or to upgrade storage (moreover for instantiable pallet). If pallet is published, make sure to warn about this breaking change.

So users of the `Purchase` pallet must be careful about the name they used in `construct_runtime!`. Hence the `runtime-migration` label, which might not be needed depending on the configuration of the `Purchase` pallet.

### CODE REVIEW

File: runtime/common/src/purchase.rs [hunk 1]
```diff
@@ -14,38 +14,19 @@
...
 // along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
 
-//! Module to process purchase of DOTs.
+//! Pallet to process purchase of DOTs.
 
 use parity_scale_codec::{Encode, Decode};
 use sp_runtime::{Permill, RuntimeDebug, DispatchResult, DispatchError, AnySignature};
 use sp_runtime::traits::{Zero, CheckedAdd, Verify, Saturating};
-use frame_support::{decl_event, decl_storage, decl_module, decl_error, ensure};
+use frame_support::{pallet_prelude::*, ensure};
```
Comment (KiChjang) [inline runtime/common/src/purchase.rs:22]:
```suggestion
use frame_support::pallet_prelude::*;
```

File: runtime/common/src/purchase.rs [hunk 2]
```diff
@@ -14,38 +14,19 @@
...
 use frame_support::traits::{
 	EnsureOrigin, Currency, ExistenceRequirement, VestingSchedule, Get
 };
-use frame_system::ensure_signed;
+use frame_system::{pallet_prelude::*, ensure_signed};
```
Comment (KiChjang) [inline runtime/common/src/purchase.rs:26]:
```suggestion
use frame_system::pallet_prelude::*;
```

File: runtime/common/src/purchase.rs [hunk 3]
```diff
@@ -143,45 +163,44 @@ decl_error! {
...
-		Statement: Vec<u8>;
-		// The block where all locked dots will unlock.
-		UnlockBlock: T::BlockNumber;
-	}
-}
+	// A map of all participants in the DOT purchase process.
+	#[pallet::storage]
+	pub(super) type Accounts<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat, T::AccountId,
```
Comment (KiChjang) [inline runtime/common/src/purchase.rs:171]:
```suggestion
		Blake2_128Concat,
		T::AccountId,
```

File: runtime/common/src/purchase.rs [hunk 4]
```diff
@@ -101,31 +82,70 @@ pub struct AccountStatus<Balance> {
...
+
+		/// The origin allowed to set account status.
+		type ValidityOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The origin allowed to make configurations to the pallet.
+		type ConfigurationOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The maximum statement length for the statement users to sign when creating an account.
+		#[pallet::constant]
+		type MaxStatementLength: Get<u32>;
```
Comment (KiChjang) [inline runtime/common/src/purchase.rs:112]: Why is this changed into a `u32`?

File: runtime/common/src/purchase.rs (hunk 3 above)
Comment (shaunxw) [reply runtime/common/src/purchase.rs:171]: imho hasher and key type in same line makes `StorageMap`/`StorageDoubleMap` more readable: it's easier to tell which is the key part and the value part.

File: runtime/common/src/purchase.rs (hunk 4 above)
Comment (shaunxw) [reply runtime/common/src/purchase.rs:112]: `MaxStatementLength` constant is `u32`. Changing the type would make it more clean, or we have to use `pallet::extra_constants` to manually convert the type. There is no breaking change so imho it's reasonable. (example of changing length type in [Substrate migration PR](https://github.com/paritytech/substrate/pull/8762/files#diff-e44afa3f58035adbf7d19adf68eed43df626e55f7fba287641e4a4546d104b18R147))

Comment (KiChjang) [reply runtime/common/src/purchase.rs:112]: Technically it is a breaking change, because `usize` on 32-bit systems is 32 bits long, but on 64-bit systems it is 64 bits long, while `u32` is fixed to be 32 bits. In other words, if the user has previously configured their `MaxStatementLength` to be larger than `u32::MAX`, they may run into issues when upgrading this pallet.

Comment (shaunxw) [reply runtime/common/src/purchase.rs:112]: By no breaking change I mean purchase is not used in any runtime yet. If `MaxStatementLength` is configured to be larger `u32::MAX`, users will have trouble running on 32-bit system even without type change here.

Comment (KiChjang) [reply runtime/common/src/purchase.rs:112]: Please do not dismiss as an ancillary concern. The issue isn't whether you can imagine cases of how it can or can't break, it's that you're changing the public API here, which IS a breaking change. I don't seem to see a good reason here other than the reluctance of using an extra attribute, unless I'm not hearing you correctly on what the difficulty is?

Comment (shaunxw) [reply runtime/common/src/purchase.rs:112]: If you insist, of course, in strict definition, it IS a breaking change.

As expressed in the beginning of this conversion, I think it will make code more clean. I have no idea why you think it's a reluctance of doing more work using an extra attribute, or other difficulty. If that's the case, I won't bother to spend time to migrate any pallet.

Comment (KiChjang) [reply runtime/common/src/purchase.rs:112]: If your cleanliness change makes a breaking API change, then I don't think it's worth it. This isn't to dismiss your work here; it's that public API changes involves a lot of other downstream changes, such as polkadot-js front-end modifications to support a different type, and others that both you and I didn't even think of at this moment in time.

Again I stress, do not dismiss public API changes as ancillary.

Comment (shaunxw) [reply runtime/common/src/purchase.rs:112]: Please don't assume I "dismiss public API changes as ancillary".

In which way exactly this type change will do breaks, other than in theory it's a breaking change? For PolkadotJS there is no new type introduced, and there is no metadata change since the constant was `u32`. This associated type is only used "publicly" in unit tests, if I don't miss anything, what is the "others that both you and I didn't even think of at this moment in time"?

Comment (KiChjang) [reply runtime/common/src/purchase.rs:112]: We're hosting this as FOSS, I don't think you can confidently say that there is absolutely 0 downstream users of this crate, especially when it's in the public domain.

Comment (shaunxw) [reply runtime/common/src/purchase.rs:112]: I could argue more about "should we do technically breaking change like this", or "how much time I've spent to fix reasonable breaking changes of substrate/polkadot since 2018".

There's just too many off-topic things happening here and sorry I don't have all day for this.  I'm happy to revert it if any party say this is making their life difficult. Or feel free to close this PR. That's it.

Comment (shawntabrizi) [reply runtime/common/src/purchase.rs:112]: This should be fine. This should be a u32 to begin with, and the value should never have exceeded `u32::max()`.

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
This pallet is not used in any runtime, and thus no worries about any migration or breaking changes here. Thank you @shaunxw

File: runtime/common/src/purchase.rs [hunk 5]
```diff
@@ -296,51 +320,57 @@ decl_module! {
...
 		/// Set the statement that must be signed for a user to participate on the DOT sale.
 		///
 		/// Origin must match the `ConfigurationOrigin`
-		#[weight = T::DbWeight::get().writes(1)]
-		fn set_statement(origin, statement: Vec<u8>) {
+		#[pallet::weight(T::DbWeight::get().writes(1))]
+		pub(super) fn set_statement(origin: OriginFor<T>, statement: Vec<u8>) -> DispatchResult {
 			T::ConfigurationOrigin::ensure_origin(origin)?;
-			ensure!(statement.len() < T::MaxStatementLength::get(), Error::<T>::InvalidStatement);
+			ensure!(statement.len() < T::MaxStatementLength::get() as usize, Error::<T>::InvalidStatement);
```
Comment (shawntabrizi) [inline runtime/common/src/purchase.rs:351]:
```suggestion
			ensure!(statement.len() as u32 < T::MaxStatementLength::get(), Error::<T>::InvalidStatement);
```

**Review by KiChjang — APPROVED** [review]

File: runtime/common/src/purchase.rs [hunk 6]
```diff
@@ -296,51 +320,57 @@ decl_module! {
...
 		/// Set the statement that must be signed for a user to participate on the DOT sale.
 		///
 		/// Origin must match the `ConfigurationOrigin`
-		#[weight = T::DbWeight::get().writes(1)]
-		fn set_statement(origin, statement: Vec<u8>) {
+		#[pallet::weight(T::DbWeight::get().writes(1))]
+		pub(super) fn set_statement(origin: OriginFor<T>, statement: Vec<u8>) -> DispatchResult {
 			T::ConfigurationOrigin::ensure_origin(origin)?;
-			ensure!(statement.len() < T::MaxStatementLength::get(), Error::<T>::InvalidStatement);
+			ensure!(statement.len() as u32 < T::MaxStatementLength::get(), Error::<T>::InvalidStatement);
```
Comment (shawntabrizi) [inline runtime/common/src/purchase.rs:351]:
```suggestion
			ensure!((statement.len() as u32) < T::MaxStatementLength::get(), Error::<T>::InvalidStatement);
```

**Review by thiolliere — APPROVED** [review]
Thanks Shawnx, the check-runtime CI is failing, but the current change doesn't affect any runtime in polkadot thus it is a false-negative

**Comment by thiolliere** [comment]
bot force merge
